README.md
# @rappopo/dab-couch-es

A [Rappopo DAB](https://github.com/rappopo/dab) implementation for CouchDB 2.0 AND Elasticsearch for finding documents. 

## Installation

Simply invoke this command in your project folder:

```
$ npm install --save @rappopo/dab-couch-es
```

And within your script:

```javascript
const DabCouchEs = require('@rappopo/dab-couch-es')
const dab = new DabCouchEs({
  couch: {
    url: 'http://localhost:5984',
    dbName: 'mydb'    
  },
  es: {
    hosts: ['localhost:9200'],
    index: 'myindex',
    type: 'mytype'
  },
  delay: 1000
})
...
dab.findOne('my-doc').then(function(doc) { ... })
```

## Options

`couch`: see [@rappopo/dab-couch](https://github.com/rappopo/dab-couch) options

`es`: see [@rappopo/dab-es](https://github.com/rappopo/dab-es) options

`delay`: time to wait (in ms) to let CouchDB fully synchronize with Elasticsearch after a create/update/remove action. Optional, defaults to 1000 ms

## Features

* [x] [find](https://docs.rappopo.com/dab/method/find/)
* [x] [findOne](https://docs.rappopo.com/dab/method/find-one/)
* [x] [create](https://docs.rappopo.com/dab/method/create/)
* [x] [update](https://docs.rappopo.com/dab/method/update/)
* [x] [remove](https://docs.rappopo.com/dab/method/remove/)
* [x] [bulkCreate](https://docs.rappopo.com/dab/method/bulk-create/)
* [x] [bulkUpdate](https://docs.rappopo.com/dab/method/bulk-update/)
* [x] [bulkRemove](https://docs.rappopo.com/dab/method/bulk-remove/)
* [x] [copyFrom](https://docs.rappopo.com/dab/method/copy-from/)
* [x] [copyTo](https://docs.rappopo.com/dab/method/copy-to/)

## Misc

* [ChangeLog](CHANGELOG.md)
* Donation: Bitcoin **16HVCkdaNMvw3YdBYGHbtt3K5bmpRmH74Y**

## License

(The MIT License)

Copyright © 2017 Ardhi Lukianto <ardhi@lukianto.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
# Changelog

## 0.0.7

* Updated to @rappopo/dab@0.0.11

## 0.0.6

* Updated to @rappopo/dab@0.0.8

## 0.0.5

* Updated to @rappopo/dab-couch@0.0.6
* Updated to @rappopo/dab-es@0.0.6
* Updated to @rappopo/dab-dab@0.0.6
* Add delay to create/update/remove

## 0.0.4

* Spec changed on bulkRemove

## 0.0.3

* Adding .asCallback to the Promise

## 0.0.2

* Adding bulkCreate method
* Adding bulkUpdate method
* Adding bulkRemove method
* Update README

## 0.0.1

* First commit ever
* Support the most basic methods:
  * find
  * findOne
  * create
  * update
  * remove

test/_lib.js
'use strict'

const fs = require('fs'),
  _ = require('lodash'),
  async = require('async'),
  elasticsearch = require('elasticsearch'),
  mapping = require('./_mapping.json')

function recreate(me, es, o, callback) {
  let body = []
  me.dummyData.forEach(function (item) {
    body.push({ index: { _id: item._id }})
    body.push(_.omit(item, ['_id']))
  })
  es.deleteByQuery({
    index: me[o].es.index,
    body: {
      query: {
        match_all: {}
      }
    }
  }, function(err, resp) {
    es.bulk({
      index: me[o].es.index,
      type: 'doc',
      refresh: 'true',
      body: body
    }, function (err, resp) {
      es.indices.refresh({
        index: me[o].es.index
      }, function (err, resp) {
        if (err) return callback(err)
        callback()                
      })
    })
  })
}

module.exports = {
  _: _,
  options: {
    couch: {
      url: 'http://localhost:5984',
      dbName: 'test',
    },
    es: {
      hosts: ['nuc:9200'],
      index: 'test'
    }
  },
  options1: {
    couch: {
      url: 'http://localhost:5984',
      dbName: 'test1',
    },
    es: {
      hosts: ['nuc:9200'],
      index: 'test1'
    }
  },
  dummyData: [
    { _id: 'jack-bauer', name: 'Jack Bauer' },
    { _id: 'james-bond', name: 'James Bond' }
  ],
  bulkDocs: [
    { _id: 'jack-bauer', name: 'Jack Bauer' },
    { _id: 'johnny-english', name: 'Johnny English' },
    { name: 'Jane Boo' }
  ],
  timeout: 5000,
  resetDb: function (callback) {
    let me = this

    async.mapSeries(['options', 'options1'], function(o, callb) {
      let es = new elasticsearch.Client(_.cloneDeep(me[o].es)),
        nano = require('nano')(me[o].couch.url)
      nano.db.destroy(me[o].couch.dbName, function (err, body) {
        if (err && err.statusCode !== 404) return callb(err)
        nano.db.create(me[o].couch.dbName, function (err, body) {
          if (err) return callb(err)
          let db = nano.db.use(me[o].couch.dbName)
          db.bulk({ docs: me.dummyData }, function (err, body) {
            if (err) return callb(err)
            // elasticsearch part
            es.indices.exists({
              index: me[o].es.index      
            }, function (err, resp) {
              if (resp) return recreate(me, es, o, callb)
              es.indices.create({
                index: me[o].es.index,
                body: mapping
              }, function (err, resp) {
                if (err) return callb(err)
                recreate(me, es, o, callb)
              })
            })
          })
        })
      })
    }, callback)

  }
}

test/1-find.js
'use strict'

const chai = require('chai'),
  chaiAsPromised = require('chai-as-promised'),
  chaiSubset = require('chai-subset'),
  expect = chai.expect

chai.use(chaiSubset)
chai.use(chaiAsPromised)

const Cls = require('../index'),
  lib = require('./_lib')

describe('find', function () {
  before(function (done) {
    this.timeout(lib.timeout)
    lib.resetDb(function (err) {
      if (err) throw err
      done()
    })
  })

  it('should return empty value', function () {
    const cls = new Cls(lib.options)
    return expect(cls.find({ query: { _id: 'no-agent' }})).to.eventually.have.property('data').that.have.lengthOf(0)
  })

  it('should return all values', function () {
    const cls = new Cls(lib.options)
    return expect(cls.find()).to.eventually.have.property('data').that.have.lengthOf(2).and.containSubset(lib.dummyData)
  })

  it('should return filtered values', function () {
    const cls = new Cls(lib.options)
    return expect(cls.find({ query: { _id: 'jack-bauer' }})).to.eventually.have.property('data').that.have.lengthOf(1).and.containSubset([lib.dummyData[0]])
  })

  it('should return 2nd page', function () {
    const cls = new Cls(lib.options)
    return expect(cls.find({ limit: 1, page: 2 })).to.eventually.have.property('data').that.containSubset([lib.dummyData[1]])
  })

  it('should sort in descending order', function () {
    const cls = new Cls(lib.options)
    return expect(cls.find({ sort: [{ _id: 'desc' }] })).to.eventually.have.property('data').that.containSubset([lib.dummyData[1]])
  })

})


test/3-create.js
'use strict'

const chai = require('chai'),
  chaiAsPromised = require('chai-as-promised'),
  expect = chai.expect

chai.use(chaiAsPromised)

const Cls = require('../index'),
  lib = require('./_lib'),
  body = {
    _id: 'jason-bourne',
    name: 'Jason Bourne'
  }

describe('create', function () {
  beforeEach(function (done) {
    this.timeout(lib.timeout)
    lib.resetDb(function (err) {
      if (err) throw err
      done()
    })
  })

  it('should return error if doc exists', function () {
    const cls = new Cls(lib.options)
    return expect(cls.create(lib.dummyData[0])).to.be.rejectedWith('Exists')
  })

  it('should return the correct value', function () {
    const cls = new Cls(lib.options)
    return expect(cls.create(body)).to.eventually.have.property('data').that.include(body)
  })
})

index.js
'use strict'

const Dab = require('@rappopo/dab'),
  DabCouch = require('@rappopo/dab-couch'),
  DabEs = require('@rappopo/dab-es')

const delay = (time) => (result) => new Promise(resolve => setTimeout(() => resolve(result), time))

class DabCouchEs extends Dab {
  constructor (options) {
    super(options)
    this.dabCouch = null
    this.dabEs = null
    this.client = {}
  }

  setOptions (options) {
    options.couch = options.couch || {}
    options.es = options.es || {}
    super.setOptions(this._.merge(this.options, {
      couch: {
        url: options.couch.url || 'http://localhost:5984',
        dbName: options.couch.dbName || 'test'
      },
      es: {
        hosts: options.es.hosts || ['localhost:9200'],
        index: options.es.index || 'test',
        type: options.es.type || 'doc',
        apiVersion: options.apiVersion,
        options: options.options || {
          refresh: 'true'
        }
      },
      delay: options.delay || 1000
    }))
  }

  setClient (params) {
    if (!this._.isEmpty(this.client)) return
    const defOptions = this._.omit(this.options, ['couch', 'es'])
    this.dabCouch = new DabCouch(this._.merge(defOptions, this.options.couch))
    this.dabEs = new DabEs(this._.merge(defOptions, this.options.es))
    this.client.dabCouch = this.dabCouch.client
    this.client.dabEs = this.dabEs.client
  }

  find (params) {
    this.setClient(params)
    return this.dabEs.find(params)
  }

  findOne (id, params) {
    this.setClient(params)
    return this.dabCouch.findOne(id, params)
  }

  create (body, params) {
    this.setClient(params)
    return this.dabCouch.create(body, params).then(delay(this.options.delay))
  }

  update (id, body, params) {
    this.setClient(params)
    return this.dabCouch.update(id, body, params).then(delay(this.options.delay))
  }

  remove (id, params) {
    this.setClient(params)
    return this.dabCouch.remove(id, params).then(delay(this.options.delay))
  }

  bulkCreate (body, params) {
    this.setClient(params)
    return this.dabCouch.bulkCreate(body, params).then(delay(this.options.delay))
  }

  bulkUpdate (body, params) {
    this.setClient(params)
    return this.dabCouch.bulkUpdate(body, params).then(delay(this.options.delay))
  }

  bulkRemove (body, params) {
    this.setClient(params)
    return this.dabCouch.bulkRemove(body, params).then(delay(this.options.delay))
  }

}

module.exports = DabCouchEs